### PR TITLE
#13405: TTNN implementation of LENET model

### PR DESCRIPTION
**Pull request by sabira-mcw**
### Ticket
[Link to the GitHub Issue](https://github.com/tenstorrent/tt-metal/issues/13405)

### Problem description
TTNN implementation for LENET model.

### What's changed
tnn support is enabled for lenet model along with demo, e2e and device test perf.

### Checklist
- [x] Post commit CI passes
- [ ] Blackhole Post commit (if applicable)
- [x] Model regression CI testing passes (if applicable)
- [x] Device performance regression CI testing passes (if applicable)
- [x] New/Existing tests provide coverage for changes

CI Links:

(Single-card) Tests for new models       - [Link](https://github.com/tenstorrent/tt-metal/actions/runs/12273855198/job/34247084545)(Lenet tests passed)
All Post Commit              - [Link](https://github.com/tenstorrent/tt-metal/actions/runs/12275562532)

### CODE REVIEW

**Comment by sabira-mcw** [comment]
@tt-rkim 

Please find below the CI links for the LeNet model.

(Single-card) Device tests : 
        - [n300](https://github.com/tenstorrent/tt-metal/actions/runs/12114720991/job/33771871030#step:9:1561)(Lenet Passed) 
        - [GS](https://github.com/tenstorrent/tt-metal/actions/runs/12114720991/job/33771870785)
All post-commit tests : 
        - [Link](https://github.com/tenstorrent/tt-metal/actions/runs/12114714045)
(Single-card) Model perf tests : 
       - [n300](https://github.com/tenstorrent/tt-metal/actions/runs/12114718853/job/33771839099#step:10:5446) 
       - [GS](https://github.com/tenstorrent/tt-metal/actions/runs/12114718853/job/33771838646#step:10:7430)(updated with the perf numbers)
(Single-card) Demo regressions :
      - [Link](https://github.com/tenstorrent/tt-metal/actions/runs/12114708958/job/33771821562#step:9:6421)
Nightly fast dispatch tests : 
      - [N150](https://github.com/tenstorrent/tt-metal/actions/runs/12114710942/job/33771826510#step:8:3725) 
      - [GS](https://github.com/tenstorrent/tt-metal/actions/runs/12114710942/job/33771825783)
      -  [N300](https://github.com/tenstorrent/tt-metal/actions/runs/12114710942/job/33771827156)

**Review by tt-rkim — APPROVED** [review]
Wonderful

**Comment by uaydonat** [comment]
@esmalTT please review the code, model, test coverage, pcc checks, etc.

**Review by uaydonat — CHANGES_REQUESTED** [review]
This will be blocked until @esmalTT complete his review.

File: models/demos/lenet/demo/demo.py
Comment (esmalTT) [inline models/demos/lenet/demo/demo.py:52]: I would argue accuracy > 1 is a bug, maybe update the code to reflect that.

File: models/demos/lenet/demo/demo.py
Comment (esmalTT) [inline models/demos/lenet/demo/demo.py:65]: If there is a reason you are disabling the kernel cache, maybe leave a comment to explain why? If the model fails with it, that is a bug and we should file an issue.

File: models/demos/lenet/lenet_utils.py
Comment (esmalTT) [inline models/demos/lenet/lenet_utils.py:43]: Should probably be `model` not `model2` since it's the only model in this scope.

File: models/demos/lenet/lenet_utils.py
Comment (esmalTT) [inline models/demos/lenet/lenet_utils.py:42]: Should probably just be `path`, users probably won't know what WEKA (and don't need to).

File: models/demos/lenet/demo/demo.py
Comment (esmalTT) [inline models/demos/lenet/demo/demo.py:25]: Is there a mistake here? I feel like you wanted to do this instead:

```python
    model = torch_LeNet.float()
    model = model.eval()
```

File: models/demos/lenet/tt/tt_lenet.py
Comment (esmalTT) [inline models/demos/lenet/tt/tt_lenet.py:1]:
```suggestion
# SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
```

File: models/demos/lenet/tests/test_perf_lenet.py
Comment (esmalTT) [inline models/demos/lenet/tests/test_perf_lenet.py:1]:
```suggestion
# SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
```

File: models/demos/lenet/lenet_utils.py
Comment (esmalTT) [inline models/demos/lenet/lenet_utils.py:1]:
```suggestion
# SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
```

File: models/demos/lenet/demo/demo.py
Comment (esmalTT) [inline models/demos/lenet/demo/demo.py:1]:
```suggestion
# SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
```

File: models/demos/lenet/tt/tt_lenet.py
Comment (esmalTT) [inline models/demos/lenet/tt/tt_lenet.py:49]: PEP8 specifies that function are `snake_case`. Consider updating this function name.

File: models/demos/lenet/tt/tt_lenet.py
Comment (esmalTT) [inline models/demos/lenet/tt/tt_lenet.py:43]: Do you want this enabled?

File: tests/ttnn/integration_tests/lenet/test_lenet.py
Comment (esmalTT) [inline tests/ttnn/integration_tests/lenet/test_lenet.py:41]: Consider increasing the PCC - I'm guessing a small model like this one would have PCC > 0.9999 if implemented correctly.

File: tests/ttnn/integration_tests/lenet/test_lenet.py
Comment (esmalTT) [inline tests/ttnn/integration_tests/lenet/test_lenet.py:1]: Newer models tend to keep their tests next to their source code. I feel like organizationally it makes the most sense to keep this in `models/demos/lenet/tests`.

@tt-rkim Any opinion on this? I know there are older models using `tests/ttnn/integration_tests`, but our team is opting for keeping model source+tests together.

**Review by esmalTT — CHANGES_REQUESTED** [review]

File: models/demos/lenet/demo/demo.py
Comment (sabira-mcw) [reply models/demos/lenet/demo/demo.py]: The kernel cache is disabled here to maintain consistency with some previous implementations I've referenced. However, the demo works fine even if this line is commented out. I’ve updated the file by removing the line.

File: models/demos/lenet/tt/tt_lenet.py
Comment (sabira-mcw) [reply models/demos/lenet/tt/tt_lenet.py:43]: The debug=True flag was enabled  primarily for internal logging during development. Since the default is False, I have updated it accordingly.

File: tests/ttnn/integration_tests/lenet/test_lenet.py
Comment (sabira-mcw) [reply tests/ttnn/integration_tests/lenet/test_lenet.py]: The PCC is 0.9993, which is very close to 0.9999. I have updated it accordingly.

**Review by esmalTT — APPROVED** [review]
Great work - code is in good shape 👍 

Before merging, please rebase, run pipelines again, and make sure they are passing to make sure everything is OK with the latest changes.

**Comment by sabira-mcw** [comment]
Thanks for the feedback! 👍 I’ll rebase the code, run the pipelines again, and ensure everything passes before merging.

**Review by tt-rkim — APPROVED** [review]

**Review by uaydonat — APPROVED** [review]

**Review by ayerofieiev-tt — APPROVED** [review]

**Comment by tt-rkim** [comment]
I have reverted this because there is an issue ttnn nightly tests

**Pull request merged**
